### DISCUSSION TITLE
Add tracing variable from middleware

### DISCUSSION THREAD

**Discussion by BenJeau** [tokio-rs/axum#1706] [Q&A] [answered]
Hi, 

I'm extracting user information in a middleware and am using the `TraceLayer`, e.g. here is the router that I'm using

```rs
Router::new()
        .nest("/api/v1", app())
        .layer(TraceLayer::new_for_http())
        .route_layer(middleware::from_fn_with_state(
            app_state.clone(),
            middleware,
        ))
```

I'm unsure how to tackle this, I've tried the following, but it was not working correctly:
* Have the following in the middleware function (I think this is not working, since the guard gets dropped at the end of the function)
```rs
let span = span!(Level::ERROR, "user", username);
let _guard = span.enter();
```

I wouldn't mind adding it to the `request` trace from the `TraceLayer`

**Comment by jplatte** [comment, marked as answer]
You need to use `.make_span_with` on the `TraceLayer` to add the custom fields to the span (see [this example](https://docs.rs/tower-http/latest/tower_http/trace/index.html#recording-fields-on-the-span)), then use `Span::current().record("field", value)` in the middleware to set its value.

**Comment by BenJeau** [comment]
Thank you for the response! I've tried what you're suggesting and it does not seem to work, seems like the middleware gets called before the `TraceLayer`, e.g.

```
2023-01-20T13:08:44.836822Z  INFO rest_server::authentication::middleware: before adding user to request
2023-01-20T13:08:44.836913Z  INFO rest_server::authentication::middleware: after adding user to request
2023-01-20T13:08:44.839237Z  INFO rest_server::api: tracing layer
```

I found out that I need it to be like the following:

```rust
Router::new()
        .nest("/api/v1", app())
        .route_layer(middleware::from_fn_with_state(
            app_state.clone(),
            middleware,
        ))
        .layer(TraceLayer::new_for_http())
```

And If I do what you suggest, everything works, thank you :)

Now I see the following:

```
2023-01-20T13:52:33.764095Z  INFO request{method=GET uri=/enrichments version=HTTP/1.1}: rest_server::authentication::middleware: before adding user to request
2023-01-20T13:52:33.764151Z  INFO request{method=GET uri=/enrichments version=HTTP/1.1 username="admin"}: rest_server::authentication::middleware: after adding user to request
```

**Reply by davidpdrsn** [reply]
See https://docs.rs/axum/latest/axum/middleware/index.html#ordering for more info about middleware ordering.

**Reply by BenJeau** [reply]
Initially thought the ordering was the other way around - thanks for the docs
